node/src/components/linear_chain_sync/operations.rs

```diff
@@ -484,19 +522,170 @@ pub(crate) async fn run_fast_sync_task(
+                continue;
+            }
+        };
+        sync_deploys_and_transfers_and_state(effect_builder, &most_recent_block).await?;
+        if let Some(key_block_info) =
+            KeyBlockInfo::maybe_from_block_header(most_recent_block.header())
+        {
+            trusted_key_block_info = key_block_info;
+        }
+    }
+
+    Ok((trusted_key_block_info, most_recent_block.take_header()))
+}
+
+/// Runs the fast synchronization task.
+pub(crate) async fn run_fast_sync_task(
+    effect_builder: EffectBuilder<JoinerEvent>,
+    trusted_hash: BlockHash,
+    chainspec: Arc<Chainspec>,
+    node_config: NodeConfig,
+) -> Result<BlockHeader, LinearChainSyncError> {
+    // Fetch the trusted header
+    let trusted_block_header = fetch_and_store_block_header(effect_builder, trusted_hash).await?;
+
+    let era_duration: TimeDiff = std::cmp::max(
+        chainspec.highway_config.min_round_length() * chainspec.core_config.minimum_era_height,
+        chainspec.core_config.era_duration,
     );
-    let mut outstanding_trie_keys = vec![Blake2bHash::from(
-        *most_recent_block_header.state_root_hash(),
-    )];
-    while let Some(trie_key) = outstanding_trie_keys.pop() {
-        let missing_descendant_trie_keys =
-            fetch_trie_and_insert_into_trie_store(effect_builder, trie_key).await?;
-        outstanding_trie_keys.extend(missing_descendant_trie_keys);
+
+    if trusted_block_header.timestamp()
+        + era_duration
+            * chainspec
+                .core_config
+                .unbonding_delay
+                .saturating_sub(chainspec.core_config.auction_delay)
+        < Timestamp::now()
+    {
+        warn!(
+            ?trusted_block_header,
+            "Timestamp of trusted hash is older than \
+             era_duration * (unbonding_delay - auction_delay)"
+        );
+    }
+
+    let maybe_last_emergency_restart_era_id = chainspec.protocol_config.last_emergency_restart;
+    match maybe_last_emergency_restart_era_id {
+        Some(last_emergency_restart_era)
+            if last_emergency_restart_era > trusted_block_header.era_id() =>
+        {
+            return Err(
+                LinearChainSyncError::TryingToJoinBeforeLastEmergencyRestartEra {
+                    last_emergency_restart_era,
+                    trusted_hash,
+                    trusted_block_header,
+                },
+            );
+        }
+        _ => {}
     }
 
+    // If we are at an upgrade:
+    // 1. Sync the trie store
+    // 2. Get the trusted era validators from the last switchblock
+    // 3. Try to get the next block by height; if there is `None` then switch to the participating
+    //    reactor.
+    if trusted_block_header.is_switch_block()
+        && trusted_block_header.era_id().successor()
+            == chainspec.protocol_config.activation_point.era_id()
+    {
+        sync_trie_store(effect_builder, *trusted_block_header.state_root_hash()).await?;
```

Comment:
Why do we have to download the trie store at the upgrade? Is that b/c Chainspec params are stored there?

Reply:
I think it's so that the participating reactor can actually apply the upgrade: That requires the pre-upgrade state. (Please correct me if I'm wrong, @xcthulhu!)

On the other hand, if the next block _does_ exist, we don't need to do this, do we? So…

Reply:
In an emergency restart, a clean-slate node might have to join the network and become a validator.  In this case, the node would need to have the global state before running upgrade.

@fizyk20 is out for a while but this is what we worked out last week.